### ISSUE TITLE
unable to moveok cancel buttons in popup

### ISSUE THREAD

**Issue by billreed63** [maqetta/maqetta#1021]
I was mocking up datagrid and got to a point where I could not move the ok button in a pop div
Using the html below:
In page designer change state to label_popup
Try to select and move the ok button in the pop up div I usally got a copy of the button
I found this on maqetta.org preview 3

``` html
<!DOCTYPE html>
<html>
<head>
<title>Untitled</title>
<style type="text/css">
@import "themes/claro/claro.css";
@import "app.css";
@import "lib/dojo/dojo/resources/dojo.css";
@import "lib/dojo/dojox/grid/resources/Grid.css";
@import "lib/dojo/dojox/grid/resources/claroGrid.css";
</style>
<script type="text/javascript" djConfig="parseOnLoad: true" src="lib/dojo/dojo/dojo.js"></script>
<script type="text/javascript">
dojo.require('dijit.layout.BorderContainer');
dojo.require('dijit.layout.ContentPane');
dojo.require('dijit.layout.TabContainer');
dojo.require('dojo.data.ItemFileReadStore');
dojo.require('dojox.grid.DataGrid');
dojo.require('dijit.form.ComboBox');
dojo.require('dijit.form.Button');
dojo.require('dojox.io.xhrScriptPlugin');</script>
<script type="text/javascript">
dojo.require('dijit.form.Button');
dojo.require('dijit.layout.TabContainer');
dojo.require('dijit.layout.ContentPane');
dojo.require('dijit.form.ComboBox');
dojo.require('dojox.io.xhrScriptPlugin');
dojo.require('dojox.grid.DataGrid');
dojo.require('dojo.data.ItemFileReadStore');
dojo.require('dijit.form.Textarea');
dojo.require('dijit.form.SimpleTextarea');</script>
<script type="text/javascript">
dojo.require('dojo.data.ItemFileReadStore');
dojo.require('dijit.form.ComboBox');
dojo.require('dojox.io.xhrScriptPlugin');
dojo.require('dojox.grid.DataGrid');
dojo.require('dijit.form.Button');
dojo.require('dijit.form.SimpleTextarea');
dojo.require('dijit.form.RadioButton');</script>
<script type="text/javascript">
dojo.require('dojo.data.ItemFileReadStore');
dojo.require('dijit.form.ComboBox');
dojo.require('dijit.form.SimpleTextarea');
dojo.require('dojox.io.xhrScriptPlugin');
dojo.require('dojox.grid.DataGrid');
dojo.require('dijit.form.Button');
dojo.require('dijit.form.RadioButton');
dojo.require('dijit.Tooltip');
dojo.require('dijit.form.CheckBox');
dojo.require('dijit.form.TextBox');</script>
<script src="maqetta/States.js"></script>
<script src="maqetta/maqetta.js"></script>
</head>
<body class="claro" dvFlowLayout="false" data-davinci-ws="collapse" dvStates="{'id_pop':{'origin':true},'label_pop':{'origin':true},'more_pop':{'origin':true},'current':'label_pop'}" id="myapp">
 <span dojoType="dojo.data.ItemFileReadStore" id="ItemFileReadStore_3" jsId="ItemFileReadStore_3" data="{&quot;identifier&quot;:&quot;uniqe_id&quot;,&quot;items&quot;:[{&quot;uniqe_id&quot;:1,&quot;id&quot;:&quot;1&quot;,&quot;label&quot;:&quot;Label 1&quot;,&quot;_more&quot;:&quot; Data&quot;},{&quot;uniqe_id&quot;:2,&quot;id&quot;:&quot;2&quot;,&quot;label&quot;:&quot;Label 2&quot;,&quot;_more&quot;:&quot; Data&quot;}]}"></span><span dojoType="dojo.data.ItemFileReadStore" id="ItemFileReadStore_4" jsId="ItemFileReadStore_4" data="{&quot;identifier&quot;:&quot;uniqe_id&quot;,&quot;items&quot;:[{&quot;uniqe_id&quot;:1,&quot;field&quot;:&quot;ID&quot;,&quot;_width&quot;:&quot;50%&quot;,&quot;_editable&quot;:&quot; &lt;select&gt;&lt;option&gt;true&lt;/option&gt;&lt;option&gt;false&lt;/option&gt;&lt;/select&gt;&quot;,&quot;_hidden&quot;:&quot; &lt;select&gt;&lt;option&gt;true&lt;/option&gt;&lt;option selected&gt;false&lt;/option&gt;&lt;/select&gt;&quot;,&quot;_type&quot;:&quot; &lt;select&gt;&lt;option&gt;Text&lt;/option&gt;&lt;option&gt;Boolean&lt;/option&gt;&lt;option&gt;Select&lt;/option&gt;&lt;/select&gt;&quot;,&quot;_options&quot;:&quot;  &quot;},{&quot;uniqe_id&quot;:2,&quot;field&quot;:&quot;Label&quot;,&quot;_width&quot;:&quot; 50% &quot;,&quot;_editable&quot;:&quot;&lt;select&gt;&lt;option&gt;true&lt;/option&gt;&lt;option&gt;false&lt;/option&gt;&lt;/select&gt;&quot;,&quot;_hidden&quot;:&quot; &lt;select&gt;&lt;option&gt;true&lt;/option&gt;&lt;option selected&gt;false&lt;/option&gt;&lt;/select&gt;&quot;,&quot;_type&quot;:&quot; &lt;select&gt;&lt;option&gt;Text&lt;/option&gt;&lt;option&gt;Boolean&lt;/option&gt;&lt;option&gt;Select&lt;/option&gt;&lt;/select&gt;&quot;}]}"></span><div style="border: none; border-radius: 6px; -moz-border-radius: 6px; border-color: #8ea8a7; border-width: 1px; border-style: solid; position: absolute; left: 214px; top: 30px; width: 611px; height: 606px;">
<select dojoType="dijit.form.ComboBox" value="Data file from workspace" disabled="false" readOnly="false" intermediateChanges="false" trim="false" uppercase="false" lowercase="false" propercase="false" required="false" style="width: 20em; position: absolute; left: 30px; top: 39px;" onclick="davinci.states.setState('csv_view')">
  <option dvwidget="html.option" value="Dummy data" selected="false">
    Dummy data</option>
  <option dvwidget="html.option" value="Comma separated data" selected="false">
    Comma separated data</option>
  <option dvwidget="html.option" value="Data file from workspace" selected="true">
    Data file from workspace</option>
  <option dvwidget="html.option" value="Data from URL (JSONP)" selected="false">
    Data from URL (JSONP)</option>
</select>
<textarea type="text" dojoType="dijit.form.SimpleTextarea" disabled="false" readOnly="false" intermediateChanges="false" rows="10" cols="10" trim="false" uppercase="false" lowercase="false" propercase="false" style="width: 510px; height: 20px; left: 30px; top: 71px; position: absolute;" value="sample_data/sample.json"></textarea>
<span dojoType="dojox.grid.DataGrid" structure="[{&quot;width&quot;:&quot;auto&quot;,&quot;name&quot;:&quot;ID&quot;,&quot;field&quot;:&quot;id&quot;},{&quot;width&quot;:&quot;auto&quot;,&quot;name&quot;:&quot;Label&quot;,&quot;field&quot;:&quot;label&quot;},{&quot;width&quot;:&quot;auto&quot;,&quot;name&quot;:&quot; More&quot;,&quot;field&quot;:&quot;_more&quot;}]" autoWidth="false" autoHeight="false" singleClickEdit="false" columnReordering="false" selectable="false" delayScroll="false" style="display: block; width: 550px; height: 178px; position: absolute; left: 27px; top: 115px;" store="ItemFileReadStore_3" escapeHTMLInData="true" dvStates="{'csv_view':{'style':{'display':'none'}}}"></span><div dvwidget="html.stickynote" style="position: absolute; height: 119px; width: 296px; background-color: #F9FFC2; border-color: black; border-width: 2px; border-style: solid; z-index: 25; -webkit-border-radius: 5px 5px 0 0; -moz-border-radius: 5px 5px 0 0; border-radius: 5px 5px 0 0; left: 610px; top: 335px;">
  <ul>
    <li>
      <i>
                Data would be read only</i>
    </li>
    <li>
      <i>
                Reorder&nbsp;columns&nbsp;by&nbsp;dragging</i>
    </li>
    <li>
      <i>
                        Reorder data by dragging rows</i>
    </li>
  </ul>
</div>
<div dvwidget="html.stickynote" style="position: absolute; height: 115px; width: 440px; background-color: #F9FFC2; border-color: black; border-width: 2px; border-style: solid; z-index: 25; -webkit-border-radius: 5px 5px 0 0; -moz-border-radius: 5px 5px 0 0; border-radius: 5px 5px 0 0; left: 627px; top: 519px;">
  <ul>
    <li>
      <i>

         Defines the&nbsp;column options.</i>
    </li>
    <li>
      <i>

         Options column is editable only when column type is set to select&nbsp;</i>
    </li>
    <li>
      <i>

Field&nbsp;column&nbsp;extracted from DataStore - Read only</i>
    </li>
    <li>
      <i>

Width column editable</i>
    </li>
  </ul>
</div>
<input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Ok" iconClass="dijitNoIcon" style="position: absolute; left: 455px; top: 560px;" class="ok_button"></input>
  <input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Cancel" iconClass="dijitNoIcon" style="position: absolute; left: 523px; top: 560px;"></input>
  <input showlabel="true" type="radio" dojoType="dijit.form.RadioButton" style="left: 26px; top: 312px; position: absolute;"></input>
  <label style="left: 52px; top: 314px; position: absolute;">

  Plan Text (label 1...</label>
  <label style="left: 52px; top: 336px; position: absolute;">

  HTML markup</label>
  <input showlabel="true" type="radio" dojoType="dijit.form.RadioButton" disabled="false" readOnly="false" intermediateChanges="false" iconClass="dijitNoIcon" checked="false" style="left: 26px; top: 334px; position: absolute;"></input>
  <img src="folder.jpg" style="left: 549px; top: 75px; position: absolute; width: 29px; height: 18px;"></img>
  <label style="left: 264px; top: 42px; position: absolute;">
    Select a file from workspace.</label>
  <img src="pencil.png" style="left: 187px; top: 119px; position: absolute; width: 16px; height: 13px;" onclick="davinci.states.setState('id_pop')"></img>
  <img src="http://maqetta.org:55556/maqetta/user/breed@nc.rr.com/ws/workspace/project2/pencil.png" style="left: 375px; top: 121px; position: absolute; width: 16px; height: 13px;" onclick="davinci.states.setState('label_pop')"></img>
  <div style="background-color: white; left: 397px; top: 119px; position: absolute; opacity: 0.6; width: 177px; height: 174px;">
</div>
  <img src="http://maqetta.org:55556/maqetta/user/breed@nc.rr.com/ws/workspace/project2/pencil.png" style="left: 554px; top: 119px; position: absolute; width: 16px; height: 13px;" onclick="davinci.states.setState('more_pop')"></img>
</div>
<div id="id_pop" style="border-radius: 6px; -moz-border-radius: 6px; border-color: #8ea8a7; border-width: 1px; border-style: solid; background-color: white; display: none; width: 168px; height: 141px; left: 95px; top: 279px; position: absolute;" dvStates="{'id_pop':{'style':{'display':'','width':'168px','height':'170px','left':'273px','top':'159px','position':'absolute'}},'label_pop':{'style':{'display':'none','left':'437px','top':'159px','position':'absolute','width':'168px','height':'227px'}}}">
<label style="position: absolute; left: 10px; top: 20px;">
Hidden</label>
<input type="checkbox" dojoType="dijit.form.CheckBox" disabled="false" readOnly="false" intermediateChanges="false" iconClass="dijitNoIcon" checked="false" style="position: absolute; left: 64px; top: 18px;"></input>
<input type="text" dojoType="dijit.form.TextBox" value="50%" disabled="false" readOnly="false" intermediateChanges="false" trim="false" uppercase="false" lowercase="false" propercase="false" style="position: absolute; left: 65px; top: 67px; width: 85px;" dvStates="{'id_pop':{'style':{'left':'64px','top':'67px','position':'absolute'}},'current':'id_pop'}"></input>
<label style="position: absolute; left: 10px; top: 67px;">
Width</label>
<input type="checkbox" dojoType="dijit.form.CheckBox" disabled="false" readOnly="false" intermediateChanges="false" iconClass="dijitNoIcon" checked="false" style="position: absolute; left: 66px; top: 41px;"></input>
<label style="position: absolute; left: 12px; top: 46px;">
Editable</label>
<label style="position: absolute; left: 12px; top: 93px;">
Type</label>
<input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Ok" iconClass="dijitNoIcon" style="left: 0px; top: 0px; position: static;" class="ok_button"></input>
<select dojoType="dijit.form.ComboBox" value="Text" disabled="false" readOnly="false" intermediateChanges="false" trim="false" uppercase="false" lowercase="false" propercase="false" required="false" style="left: 66px; top: 94px; position: absolute; width: 80px;" dvStates="{'current':'label_pop'}">
<option dvwidget="html.option" value="Text">
Text</option>
<option dvwidget="html.option" value="Boolean">
Boolean</option>
<option dvwidget="html.option" value="Select">
Select</option>
</select>
<input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Ok" iconClass="dijitNoIcon" style="left: 0px; top: 0px; position: static;" class="ok_button" dvStates="{'id_pop':{'style':{'left':'36px','top':'128px','position':'absolute'}},'current':'id_pop'}"></input>
</div>
<div id="label_pop" style="border-radius: 6px; -moz-border-radius: 6px; border-color: #8ea8a7; border-width: 1px; border-style: solid; background-color: white; display: none; width: 168px; height: 141px; left: 95px; top: 279px; position: absolute;" dvStates="{'id_pop':{'style':{'display':'none','width':'168px','height':'153px','left':'0px','top':'-6px','position':'absolute'}},'label_pop':{'style':{'display':'','left':'437px','top':'159px','position':'absolute','width':'168px','height':'227px'}}}">
<label style="position: absolute; left: 10px; top: 20px;">
Hidden</label>
<input type="checkbox" dojoType="dijit.form.CheckBox" disabled="false" readOnly="false" intermediateChanges="false" iconClass="dijitNoIcon" checked="false" style="position: absolute; left: 64px; top: 18px;"></input>
<input type="text" dojoType="dijit.form.TextBox" value="50%" disabled="false" readOnly="false" intermediateChanges="false" trim="false" uppercase="false" lowercase="false" propercase="false" style="position: absolute; left: 65px; top: 67px; width: 85px;"></input>
<label style="position: absolute; left: 10px; top: 67px;">
Width</label>
<input type="checkbox" dojoType="dijit.form.CheckBox" disabled="false" readOnly="false" intermediateChanges="false" iconClass="dijitNoIcon" checked="false" style="position: absolute; left: 66px; top: 41px;"></input>
<label style="position: absolute; left: 12px; top: 46px;">
Editable</label>
<label style="position: absolute; left: 12px; top: 93px;">
Type</label>
<select dojoType="dijit.form.ComboBox" value="Select" disabled="false" readOnly="false" intermediateChanges="false" trim="false" uppercase="false" lowercase="false" propercase="false" required="false" style="left: 66px; top: 94px; position: absolute; width: 80px;" dvStates="{'current':'label_pop'}">
<option dvwidget="html.option" value="Text" selected="false">
Text</option>
<option dvwidget="html.option" value="Boolean" selected="false">
Boolean</option>
<option dvwidget="html.option" value="Select" selected="true">
Select</option>
</select>
<label style="left: 12px; top: 93px; position: absolute;" dvStates="{'label_pop':{'style':{'left':'14px','top':'119px','position':'absolute'}}}">
Options</label>
<textarea type="text" dojoType="dijit.form.SimpleTextarea" rows="10" cols="10" style="left: 0px; top: 0px; position: static; width: 200px; height: 50px;" dvStates="{'label_pop':{'style':{'left':'62px','top':'122px','position':'absolute','width':'78px'}}}"></textarea>
<input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Ok" iconClass="dijitNoIcon" style="left: 0px; top: 0px; position: static;" class="ok_button" dvStates="{'label_pop':{'style':{'left':'36px','top':'191px','position':'absolute'}},'current':'label_pop'}"></input>
<input type="button" dojoType="dijit.form.Button" disabled="false" intermediateChanges="false" label="Cancel" iconClass="dijitNoIcon" style="left: 0px; top: 0px; position: static;" dvStates="{'label_pop':{'style':{'left':'104px','top':'192px','position':'absolute'}},'current':'label_pop'}"></input>
</div>
</body>
</html>

```

**Comment by JonFerraiolo** [comment]
I assigned this to myself for M4 because it relates to MoveCommand and states (see #909), although this one is likely about ReparentCommand, but Reparent and MoveCommand tend to have the same problems. Maybe @peller might end up with this one ultimately.

**Comment by JonFerraiolo** [comment]
Pretty sure this has something to do with the fact that the widget's ID has been stuffed into a subnode rather than the root node. Here is a copy/paste from the runtime state of the widget. Notice that widgetid is "input_12" but Dijit puts the ID on an interior subnode:

```
<span class="dijit dijitReset dijitInline dijitButton ok_button dijitButtonDisabled dijitDisabled" role="presentation" style="left: 0px; top: 0px; position: static; " widgetid="input_12"><span class="dijitReset dijitInline dijitButtonNode" data-dojo-attach-event="ondijitclick:_onClick" role="presentation">
    <span class="dijitReset dijitStretch dijitButtonContents HtmlWidget" data-dojo-attach-point="titleNode,focusNode" role="button" aria-labelledby="input_12_label" tabindex="-1" aria-disabled="true" id="input_12" style="-webkit-user-select: none; " properties="[object Object]">
```

**Comment by JonFerraiolo** [comment]
Looks like the problem is the combination of the facts that (a) Button puts the ID on a subnode, and (b) somehow or other that subnode has been given a _dvWidget property. I suspect the real problem here is upstream due to (b).

**Comment by JonFerraiolo** [comment]
I have discovered where the bug is. Only happens with the combination of using states and widgets who put their IDs on a subnode. What happens right now is that the restoreStates logic in Context.js (invoked at doc load time) finds the widget's node by a simple getElementById(), which of course returns the subnode rather than the widget's root node. Then when getWidget() is called on the next line, which has the side effect of turning the subnode into a dvWidget (HTMLWidget).

Current logic:

```
            var node = id == "body" ? this.getContainerNode() : this.getDocument().getElementById(id);
            var widget = davinci.ve.widget.getWidget(node);
```

Here is some candidate replacement logic:

```
            var node = id == "body" ? this.getContainerNode() : davinci.ve.widget.byId(id, this.getDocument());
            var widget = davinci.ve.widget.getWidget(node);
```

However, when I tried the candidate replacement logic, the states palette no longer switches states. (Disappointed!) Maybe there is a combination of bugs here that sometimes cancel each other out?
